controllers/configurationpolicy_controller.go

```diff
@@ -130,247 +182,188 @@ type ConfigurationPolicyReconciler struct {
 
-		duration := time.Now().UTC().Sub(before)
-		seconds := float64(duration) / float64(time.Second)
-
-		policyEvalSecondsCounter.WithLabelValues(policy.Name).Add(seconds)
-		policyEvalCounter.WithLabelValues(policy.Name).Inc()
-	}
-}
+			return reconcile.Result{}, nil
+		}
 
-// refreshDiscoveryInfo tries to discover all the available APIs on the cluster, and update the
-// cached information. If it encounters an error, it may update the cache with partial results,
-// if those seem "better" than what's in the current cache.
-func (r *ConfigurationPolicyReconciler) refreshDiscoveryInfo() error {
-	log.V(2).Info("Refreshing the discovery info")
-	r.lock.Lock()
-	defer func() { r.lock.Unlock() }()
+		requeueAfter, err = policy.Spec.EvaluationInterval.GetCompliantInterval()
+	} else {
+		// If the policy is not compliant (i.e. noncompliant or unknown), fall back to the noncompliant evaluation
+		// interval. This is a court of guilty until proven innocent.
+		if policy.Spec.EvaluationInterval.IsWatchForNonCompliant() {
+			log.V(2).Info(
+				"The policy is not compliant and has the evaluation interval set to watch. Will not schedule.",
+			)
 
-	dd := r.TargetK8sClient.Discovery()
+			return reconcile.Result{}, nil
+		}
 
-	serverVersion, err := dd.ServerVersion()
-	if err != nil {
-		log.Error(err, "Could not get the server version")
+		requeueAfter, err = policy.Spec.EvaluationInterval.GetNonCompliantInterval()
 	}
 
-	r.serverVersion = serverVersion.String()
-
-	_, apiResourceList, resourceErr := dd.ServerGroupsAndResources()
-	if resourceErr != nil {
-		log.Error(resourceErr, "Could not get the full API resource list")
+	// At this point, we know the evaluation interval isn't set to watch so remove any potential watches for this
+	// policy.
+	removeWatcherErr := r.DynamicWatcher.RemoveWatcher(policy.ObjectIdentifier())
+	if removeWatcherErr != nil {
+		log.Error(err, "Failed to remove any watches related to this ConfigurationPolicy. Will ignore.")
 	}
 
-	if resourceErr == nil || (len(apiResourceList) > len(r.discoveryInfo.apiResourceList)) {
-		// update the list if it's complete, or if it's "better" than the old one
-		r.discoveryInfo.apiResourceList = apiResourceList
-	}
+	if err != nil {
+		if errors.Is(err, policyv1.ErrIsNever) {
+			log.V(2).Info(
+				"The policy will not be scheduled for evaluation since it has an evaluation interval of never",
+			)
 
-	apiGroups, groupErr := restmapper.GetAPIGroupResources(dd)
-	if groupErr != nil {
-		log.Error(groupErr, "Could not get the full API groups list")
-	}
+			return reconcile.Result{}, nil
+		}
 
-	if (resourceErr == nil && groupErr == nil) || (len(apiGroups) > len(r.discoveryInfo.apiGroups)) {
-		// update the list if it's complete, or if it's "better" than the old one
-		r.discoveryInfo.apiGroups = apiGroups
+		log.Error(
+			err,
+			"The policy has an invalid evaluation interval; defaulting to watch",
```

Comment:
But hasn't it already removed the watches? In this case does it constantly requeue immediately?

Reply:
Good point! Fixing.